include/circt/Dialect/Verif/VerifOps.td

```diff
@@ -260,4 +263,197 @@ def YieldOp : VerifOp<"yield", [
+    assumptions. The result type must be explicitly marked. The resulting value 
+    is a new non-deterministic value for every implicit clock cycle. 
+  }];
+  let results = (outs AnyType:$result);
+
+  let assemblyFormat = "attr-dict `:` type($result)";
+}
+
+def ConcreteInputOp : VerifOp<"concrete_input", [
+  AllTypesMatch<["init", "update"]>, HasParent<"verif::FormalOp">
+]> {
+  let summary = "declare a concrete input for formal verification";
+  let description = [{
+    This operation declares a concrete input that can then be used in the 
+    reasoning surrounding symbolic inputs, allowing for a form of concolic 
+    reasoning to take place in a `verif.formal` block. Concrete inputs are 
+    defined by an initial value and an update SSA value. This is equivalent 
+    to a register with an implicit clock, which is initialized with the init 
+    value and obtains a new value every implicit clock tick from the update 
+    value.
+  }];
+
+  let arguments = (ins AnyType:$init, AnyType:$update);
+  let results = (outs AnyType:$result);
+
+  let assemblyFormat = [{
+    $init `,` $update attr-dict `:` 
+    functional-type(operands, results)
+  }];
+}
+
+//===----------------------------------------------------------------------===//
+// Formal Contract Ops
+//===----------------------------------------------------------------------===//
+
+def ContractOp : VerifOp<"contract", [Pure, HasParent<"hw::HWModuleOp">]>{
+  let summary = "declare formal contract";
+  let description = [{
+    This operation declares a formal contract that is used to create precondition 
+    and postconditions on a parent module. These are used as an abstraction to 
+    better modularize formal verification such that each module containing a contract 
+    is checked exactly once.
+          
+    e.g. 
+    ```
+    hw.module @Bar(in %clk : !seq.clock, in %foo : i8, out "" : i8, out "1" : i8) { 
+      verif.contract { 
+        %c0_8 = hw.constant 0 : i8 
+        %prec = comb.icmp bin neq %c0_8, %foo : i8
+        verif.require %precond : i1
+
+        %out0, %out1 = verif.result : i8  
+        %post = comb.icmp bin ugt %out0, %foo : i8
+        %post1 = comb.icmp bin ult %out1, %foo : i8
+        verif.ensure %post : i1
+        verif.ensure %post1 : i1
+      }
+      /* ... Module definition ... */
+    }
+    ```
+
+    This later is used to replace any instance of Bar during verification:  
+    ```
+    %bar.0, %bar.1 = hw.instance "bar" @Bar("clk": %clk : !seq.clock, "foo" : %foo : i8)  -> ("" : i8, "1" : i8)
+
+    /* ... After PrepareForFormal Pass becomes ... */
+
+    %c0_8 = hw.constant 0 : i8 
+    %prec = comb.icmp bin neq %c0_8, %bar : i8
+    verif.assert %precond : i1
+
+    %out0 = verif.symbolic_input : i8
+    %out1 = verif.symbolic_input : i8
+    %post = comb.icmp bin ugt %out0, %bar : i8
+    %post1 = comb.icmp bin ult %out1, %bar : i8
+    verif.assume %post : i1
+    verif.assume %post1 : i1
```

Comment:
I think you swapped %foo for %bar between here and the contract above?
```suggestion
    %c0_8 = hw.constant 0 : i8 
    %prec = comb.icmp bin neq %c0_8, %foo : i8
    verif.assert %precond : i1

    %out0 = verif.symbolic_input : i8
    %out1 = verif.symbolic_input : i8
    %post = comb.icmp bin ugt %out0, %foo : i8
    %post1 = comb.icmp bin ult %out1, %foo : i8
    verif.assume %post : i1
    verif.assume %post1 : i1
```

Also, I'm guessing %out0 and %out1 replace all uses of %bar.0 and %bar.1? It could be clearer to just name them %bar.0 and %bar.1 in the after PrepareForFormal section, as that implies the replacement?